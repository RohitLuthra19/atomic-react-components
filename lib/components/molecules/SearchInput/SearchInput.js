// @flow
/**
 *
 * SearchInput
 *
 */
import type { Node } from 'react';

import React from 'react';
import styled from 'styled-components';
import classnames from 'classnames';

import styles from './SearchInput.style';
import type { Props, SpeechRecognition } from './types';
import Input from '../../atoms/Input';
import Button from '../../atoms/Button';

export const createSearchInput = (global: any) => {
  const BrowserSpeechRecognition: any =
    typeof global !== 'undefined' &&
    (global.SpeechRecognition ||
      global.webkitSpeechRecognition ||
      global.mozSpeechRecognition ||
      global.msSpeechRecognition ||
      global.oSpeechRecognition);
  const speechRecognition: SpeechRecognition =
    BrowserSpeechRecognition && new BrowserSpeechRecognition();

  const browserSupportsSpeechRecognition = BrowserSpeechRecognition !== undefined;
  const SearchInput = ({
    className,
    onSearch,
    onSpeechStart,
    onSpeechEnd,
    inputProps = {},
    searchButtonProps = {},
    speechButtonProps = {},
    lang,
  }: Props): Node => {
    const [inputValue, setInputValue] = React.useState(inputProps.value || '');
    const [isListening, setIsListening] = React.useState(false);

    const setSpeechResult = event => {
      const value = (event && event.results && event.results[0][0].transcript) || '';
      if (typeof onSpeechEnd === 'function') {
        onSpeechEnd(value);
      }
      if (value) {
        setInputValue(value);
      }
      setIsListening(false);
    };

    if (browserSupportsSpeechRecognition) {
      speechRecognition.onresult = setSpeechResult;
      speechRecognition.onend = setSpeechResult;
      speechRecognition.lang = lang;
    }

    const onSearchClick = () => {
      if (typeof onSearch === 'function') {
        onSearch(inputValue);
      }
    };

    const onSpeechClick = () => {
      if (isListening) {
        speechRecognition.abort();
        setIsListening(false);
        return;
      }
      if (typeof onSpeechStart === 'function') {
        setIsListening(true);
        onSpeechStart();
      }
      speechRecognition.start();
    };

    const onInputChange = (e: SyntheticEvent<HTMLInputElement>) => {
      if (typeof inputProps.onChange === 'function') {
        inputProps.onChange(e);
      }
      setInputValue((e.currentTarget: HTMLInputElement).value);
    };

    return (
      <div className={classnames('search', isListening ? 'search--listening' : '', className)}>
        <Input
          {...inputProps}
          className={classnames('search__input', inputProps.className)}
          onChange={onInputChange}
          value={inputValue}
        />
        {browserSupportsSpeechRecognition && (
          <Button
            {...speechButtonProps}
            className={classnames('search__speech-btn', speechButtonProps.className)}
            onClick={onSpeechClick}
          />
        )}
        <Button
          {...searchButtonProps}
          className={classnames('search__search-btn', searchButtonProps.className)}
          onClick={onSearchClick}
        />
      </div>
    );
  };

  return SearchInput;
};
const SearchInput = createSearchInput(window);
export default styled(SearchInput)`
  ${styles};
`;

export { SearchInput as SearchInputVanilla };
