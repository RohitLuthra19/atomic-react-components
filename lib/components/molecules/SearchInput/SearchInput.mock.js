import { action } from '@storybook/addon-actions';

const defaultConfig = {
  onSpeechStart: action('On Speech Start fired'),
  onSpeechEnd: action('On Speech End fired'),
  onSearch: action('On Search fired'),
  inputProps: {
    onBlur: action('On Blur fired'),
    onChange: action('On Change fired'),
  },
  searchButtonProps: {
    children: 'Search',
  },
  speechButtonProps: {
    children: 'Speak',
  },
};

export const stylesConfig = {
  onSpeechStart: action('On Speech Start fired'),
  onSpeechEnd: action('On Speech End fired'),
  onSearch: action('On Search fired'),
  inheritedStyles: `border: 1px solid #dfe1e5;
  height: 39px;
  height: 44px;
  &.search--listening {
    border: solid 2px red;
  }`,
  inputProps: {
    onBlur: action('On Blur fired'),
    onChange: action('On Change fired'),
    inheritedStyles: `color: blue`,
  },
  searchButtonProps: {
    inheritedStyles: `background: url(images/search.svg) 0 0 no-repeat;
    background-size: 24px;
    height: 24px;
    width: 24px;
    vertical-align: middle;
    padding: 0;
    margin: 10px;`,
  },
  speechButtonProps: {
    inheritedStyles: `background: url(images/speech.png) 0 0 no-repeat;
    background-size: 24px;
    height: 24px;
    width: 24px;
    vertical-align: middle;
    padding: 0;
    margin: 10px;`,
  },
};

export default defaultConfig;
