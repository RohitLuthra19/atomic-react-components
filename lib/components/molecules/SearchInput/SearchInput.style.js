import { css } from 'styled-components';

export default css`
  display: flex;
  ${props => (props.inheritedStyles ? props.inheritedStyles : '')};
  .search__speech-btn {
    ${props => (props.speechButtonStyles ? props.speechButtonStyles : '')};
  }

  .search__search-btn {
    ${props => (props.searchButtonStyles ? props.searchButtonStyles : '')};
  }
`;
