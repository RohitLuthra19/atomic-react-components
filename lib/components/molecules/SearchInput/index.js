// @flow
export { default } from './SearchInput';
export { SearchInputVanilla, createSearchInput } from './SearchInput';
