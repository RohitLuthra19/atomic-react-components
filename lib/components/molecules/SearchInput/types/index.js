// @flow
import type { Node } from 'react';
import type { Props as InputProps } from '../../../atoms/Input';
import type { Props as ButtonProps } from '../../../atoms/Button';

export type Props = {
  children: Function,
  className?: string,
  onSearch: Function,
  onSpeechStart: Function,
  onSpeechEnd: Function,
  inputProps?: InputProps,
  searchButtonProps?: ButtonProps,
  speechButtonProps?: ButtonProps,
  lang?: String,
  showInterimResults: boolean,
};

export type SpeechRecognition = {
  onresult: Function,
  onend: Function,
  abort: Function,
  stop: Function,
  start: Function,
  lang?: String,
  interimResults?: boolean,
};
