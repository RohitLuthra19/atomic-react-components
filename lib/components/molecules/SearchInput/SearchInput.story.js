import React from 'react';
import { storiesOf } from '@storybook/react';
// import { action } from '@storybook/addon-actions';
import defaultConfig, { stylesConfig } from './SearchInput.mock';

// Import Styled Component to showcase variations
import SearchInput, { SearchInputVanilla } from '.';

// $FlowFixMe eslint-disable-next-line extra-rules/no-commented-out-code
storiesOf('Molecules', module).addWithChapters('SearchInput', {
  chapters: [
    {
      sections: [
        {
          sectionFn: () => (
            <SearchInputVanilla
              {...defaultConfig}
              className={`hide-default-sample ${defaultConfig.className}`}
            />
          ),
          options: {
            showSource: true,
            allowSourceToggling: true,
            showPropTables: true,
            allowPropTablesToggling: true,
          },
        },
      ],
    },
    {
      title: 'Primary SearchInput Variations',
      sections: [
        {
          title: 'Default View',
          sectionFn: () => <SearchInput {...defaultConfig} />,
        },
      ],
    },
    {
      title: 'SearchInput with Custom Styles',
      sections: [
        {
          title: 'Input with Custom Styles',
          sectionFn: () => <SearchInput {...stylesConfig} />,
        },
      ],
    },
  ],
});
