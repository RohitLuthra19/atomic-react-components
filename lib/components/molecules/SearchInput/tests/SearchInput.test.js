import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';

import SearchInput, { createSearchInput } from '../index';

const createMockSearchInput = () => {
  const speechInstance = {
    start: jest.fn(),
    abort: jest.fn(),
  };
  const mockWindow = {
    SpeechRecognition: function SpeechRecognition() {
      return speechInstance;
    },
  };
  const SearchInputComp = createSearchInput(mockWindow);
  return { SearchInputComp, speechInstance };
};

describe('<SearchInput />', () => {
  let SearchInputComponent = '';

  test('should render correctly', () => {
    SearchInputComponent = shallow(<SearchInput>Test</SearchInput>);
    expect(SearchInputComponent).toMatchSnapshot();
  });

  test('calls input onchange prop on input change', () => {
    const mockOnChange = jest.fn();
    const mockEvent = { currentTarget: { value: 'Hello' } };
    const { SearchInputComp } = createMockSearchInput();
    const wrapper = mount(<SearchInputComp {...{ inputProps: { onChange: mockOnChange } }} />);
    const input = wrapper.find('.search__input').at(0);
    act(() => {
      input.simulate('change', mockEvent);
    });
    expect(mockOnChange).toHaveBeenCalled();
  });

  test('simulates search click', () => {
    const onSearch = jest.fn();
    const wrapper = mount(<SearchInput onSearch={onSearch} />);
    wrapper
      .find('.search__search-btn')
      .at(0)
      .simulate('click');
    expect(onSearch).toHaveBeenCalled();
  });

  test('onSpeechStart should be called on speech button click', () => {
    const onSpeechStart = jest.fn();
    const { SearchInputComp } = createMockSearchInput();
    const wrapper = mount(<SearchInputComp onSpeechStart={onSpeechStart} />);
    wrapper
      .find('.search__speech-btn')
      .at(0)
      .simulate('click');
    expect(onSpeechStart).toHaveBeenCalled();
  });

  test('calls setState with speech transcript on speechResult', () => {
    const mockSetInputValue = jest.fn();
    const onSpeechEnd = jest.fn();
    React.useState = jest.fn(() => ['', mockSetInputValue]);
    const mockSpeechEvent = {
      results: [[{ transcript: 'transcript' }]],
    };
    mockSpeechEvent.results[0].isFinal = true;
    const { speechInstance, SearchInputComp } = createMockSearchInput();
    mount(<SearchInputComp onSpeechEnd={onSpeechEnd} />);
    act(() => {
      speechInstance.onresult(mockSpeechEvent);
    });
    expect(mockSetInputValue).toHaveBeenCalledWith('transcript');
    speechInstance.onend();
    expect(onSpeechEnd).toHaveBeenCalled();
  });

  test('does not render speech button if speech not supported', () => {
    const mockWindow = {
      SpeechRecognition: undefined,
    };
    const SearchInputComp = createSearchInput(mockWindow);
    const wrapper = mount(<SearchInputComp />);
    expect(wrapper.find('.search__speech-btn').length).toBe(0);
  });

  test('aborts speech on clicking speech button when already listening', () => {
    const { SearchInputComp, speechInstance } = createMockSearchInput();
    // mock to set isListening to true
    React.useState = jest.fn(() => [true, jest.fn()]);

    const wrapper = mount(<SearchInputComp />);
    wrapper
      .find('.search__speech-btn')
      .at(0)
      .simulate('click');
    expect(speechInstance.abort).toHaveBeenCalled();
  });

  test('Sets language passed in props to speech instance', () => {
    const { SearchInputComp, speechInstance } = createMockSearchInput();
    // mock to set isListening to true
    const language = 'en';
    mount(<SearchInputComp lang={language} />);
    expect(speechInstance.lang).toBe(language);
  });
});
