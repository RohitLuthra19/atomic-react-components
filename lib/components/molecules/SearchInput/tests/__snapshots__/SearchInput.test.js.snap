// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> should render correctly 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "SearchInput-sc-6c1zjb-0",
        "isStatic": false,
        "rules": Array [
          "display:flex;",
          [Function],
          ";.search__speech-btn{",
          [Function],
          ";}.search__search-btn{",
          [Function],
          ";}",
          ";",
        ],
      },
      "displayName": "SearchInput",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "SearchInput-sc-6c1zjb-0",
      "target": [Function],
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
>
  Test
</StyledComponent>
`;
